src/Cli/dotnet/commands/dotnet-run/VirtualProjectBuildingCommand.cs

```diff
@@ -139,86 +178,443 @@ private ProjectInstance CreateProjectInstance(
         {
             GlobalProperties = globalProperties,
         });
+
+        ProjectRootElement CreateProjectRootElement(ProjectCollection projectCollection)
+        {
+            Debug.Assert(!_directives.IsDefault && _targetFilePath is not null, $"{nameof(PrepareProjectInstance)} should have been called first.");
+
+            var projectFileFullPath = Path.ChangeExtension(EntryPointFileFullPath, ".csproj");
+            var projectFileWriter = new StringWriter();
+            WriteProjectFile(projectFileWriter, _directives, virtualProjectFile: true, targetFilePath: _targetFilePath);
+            var projectFileText = projectFileWriter.ToString();
+
+            using var reader = new StringReader(projectFileText);
+            using var xmlReader = XmlReader.Create(reader);
+            var projectRoot = ProjectRootElement.Create(xmlReader, projectCollection);
+            projectRoot.FullPath = projectFileFullPath;
+            return projectRoot;
+        }
     }
 
-    // Kept in sync with the default `dotnet new console` project file (enforced by `DotnetProjectAddTests.SameAsTemplate`).
-    private const string CommonProjectProperties = """
-            <OutputType>Exe</OutputType>
-            <TargetFramework>net10.0</TargetFramework>
-            <ImplicitUsings>enable</ImplicitUsings>
-            <Nullable>enable</Nullable>
-        """;
+    public static void WriteProjectFile(TextWriter writer, ImmutableArray<CSharpDirective> directives)
+    {
+        WriteProjectFile(writer, directives, virtualProjectFile: false, targetFilePath: null);
+    }
 
-    public static string GetNonVirtualProjectFileText()
+    private static void WriteProjectFile(TextWriter writer, ImmutableArray<CSharpDirective> directives, bool virtualProjectFile, string? targetFilePath)
     {
-        return $"""
-            <Project Sdk="Microsoft.NET.Sdk">
+        int processedDirectives = 0;
 
+        var sdkDirectives = directives.OfType<CSharpDirective.Sdk>();
+        var propertyDirectives = directives.OfType<CSharpDirective.Property>();
+        var packageDirectives = directives.OfType<CSharpDirective.Package>();
+
+        string sdkValue = "Microsoft.NET.Sdk";
+
+        if (sdkDirectives.FirstOrDefault() is { } firstSdk)
+        {
+            sdkValue = firstSdk.ToSlashDelimitedString();
+            processedDirectives++;
+        }
+
+        if (virtualProjectFile)
+        {
+            writer.WriteLine($"""
+                <Project>
+
+                  <!-- We need to explicitly import Sdk props/targets so we can override the targets below. -->
+                  <Import Project="Sdk.props" Sdk="{EscapeValue(sdkValue)}" />
+                """);
+        }
+        else
+        {
+            writer.WriteLine($"""
+                <Project Sdk="{EscapeValue(sdkValue)}">
+
+                """);
+        }
+
+        foreach (var sdk in sdkDirectives.Skip(1))
```

Comment:
Would it be bad design-wise to just say that you can't have multiple sdks in a file? I can't think of a normal use case where you shouldn't just upgrade to a project file